src/analyze/flavors/lwc/utils.ts

```diff
@@ -0,0 +1,142 @@
+//import { Node, ClassDeclaration, getCombinedModifierFlags, ModifierFlags, SyntaxKind } from "typescript";
+import { Node, ClassDeclaration, getCombinedModifierFlags, ModifierFlags } from "typescript";
+import { AnalyzerVisitContext } from "../../analyzer-visit-context";
+import { camelToDashCase } from "../../util/text-util";
+import { existsSync, readFileSync } from "fs";
+import { parseJsDocForNode } from "../js-doc/parse-js-doc-for-node";
+
+// Discover TS nodes
+// https://ts-ast-viewer.com/#
+
+type ComponentRef = { tagName: string };
+
+const LWCCACHE = Symbol("LWC Compoeent");
+export function isLwcComponent(node: Node, context: AnalyzerVisitContext): ComponentRef | undefined {
+	if ((node as any)[LWCCACHE]) {
+		return (node as any)[LWCCACHE];
+	}
+	const r = _isLwcComponent(node, context);
+	if (r) {
+		(node as any)[LWCCACHE] = r;
+	}
+	return r;
+}
+
+function _isLwcComponent(node: Node, context: AnalyzerVisitContext): ComponentRef | undefined {
+	const { ts } = context;
+	if (ts.isClassDeclaration(node)) {
+		const jsName = node.getSourceFile().fileName;
+
+		const splitjsName = jsName.split("/");
+		if (splitjsName.length < 3) {
+			return;
+		}
+		const nameSpace = splitjsName[splitjsName.length - 3];
+		const componentName = splitjsName[splitjsName.length - 2];
+		const tagName = nameSpace + "-" + camelToDashCase(componentName);
+
+		// Main case (~100% of the cases)
+		// The class is a default export and there is a template (.html) starting with <template>
+		// Moreover the JS file name should match the directory name, minus the extension (js|ts)
+		//    https://lwc.dev/guide/reference#html-file
+		const flags = getCombinedModifierFlags(node);
+		if (flags & ModifierFlags.ExportDefault && (jsName.endsWith(".js") || jsName.endsWith(".ts"))) {
+			const fileName = splitjsName[splitjsName.length - 1];
+			const fileNoExt = fileName.substring(0, fileName.length - 3);
+			if (fileNoExt === componentName) {
+				const htmlName = jsName.substring(0, jsName.length - 3) + ".html";
+				if (existsSync(htmlName)) {
+					const content = readFileSync(htmlName, "utf8").trim();
+					if (content.startsWith("<template>")) {
+						return { tagName };
+					}
+				}
```

Comment:
When testing this on a project with non-LWC-components, it results in this code being run a couple of times because of the `my-component/my-component.ts` convention. I fear that this might result in performance regressions for users not using LWC. It is possible to move up the `inheritFromLightning`-check, so that the `_isLwcComponent` function returns right away if `LightningElement` is not inherited?

Reply:
Quick answer: checking `LightningElement` is not reliable as the inheritance can be indirect and this is currently not yet supported by the framework. On the other hand, the file content is only loaded if it exists with the same name than the component. So I would believe that the impact is light for other frameworks.
BTW, having real projects that are mixing and matching different technologies should be very rare. One great addition would be a parameter that specifies the flavors to load. This way, one can avoid unnecessary checks.

Reply:
The best solution would effectively be an option to the CLI, so you don't run the flavors you don't need. I'm actually using the CLI, so switching to a script is possible but a bit of work.
That being said, I don't believe that the file access is an actual performance regression. I just ran the tests with the profiler enabled, and the calls to `existsSync()` are insignificant on my Mac. To run the profiler, I added the code bellow to `ts-test.ts`, and then loaded the snapshots in the Chrome performance dev tools.
```typescript
import fs from 'fs';
import inspector from 'inspector';
let session = new inspector.Session();

test.before(async () => {
	session = new inspector.Session();
	session.connect();
	session.post('Profiler.enable', () => {
		session.post('Profiler.start', () => {
		});
	});
});

test.after(async () => {
	session.post('Profiler.stop', (err, { profile }) => {
		if (!err) {
		  fs.writeFileSync('./profile.cpuprofile'+(new Date()).getTime(), JSON.stringify(profile));
		}
   });
});
```
None of the profiler snapshots show a significant amount of time is `isLwcComponent()` calling `existsSync()`. The later is also called more time from other call stacks.

Reply:
I also modified how the cache is filled, so the code that discovers LWC classes is only executed once even when called multiple times.